File: common/src/main/java/com/ssblur/scriptor/api/ScriptorRegistry.java
package com.ssblur.scriptor.api;

import com.ssblur.scriptor.registry.words.WordRegistry;
import com.ssblur.scriptor.api.word.Word;
import com.ssblur.scriptor.api.word.Action;
import com.ssblur.scriptor.api.word.Descriptor;
import com.ssblur.scriptor.api.word.Subject;
import net.minecraft.resources.ResourceLocation;

public class ScriptorRegistry {
  /**
   * Registers a Word to Scriptor's WordRegistry, adding it to dictionaries.
   * @param key The key to register this word to.
   *            Will overwrite any word of the same type already written to this key.
   * @param word The word to register.
   *             Supports Actions, Subjects, and Descriptors.
   * @see Action
   * @see Subject
   * @see Descriptor
   * @return The word passed in.
   *         For chaining and static assignment.
   */
  public static Word register(String key, Word word) {
    if(word instanceof Action action) return WordRegistry.INSTANCE.register(key, action);
    if(word instanceof Subject subject) return WordRegistry.INSTANCE.register(key, subject);
    if(word instanceof Descriptor descriptor) return WordRegistry.INSTANCE.register(key, descriptor);
    throw new RuntimeException("Unknown Word type! Expected one of Action, Subject, or Descriptor.");
  }

  /**
   * Registers a Word to Scriptor's WordRegistry, adding it to dictionaries.
   * @param key The key to register this word to.
   *            Will overwrite any word of the same type already written to this key.
   * @param word The word to register.
   *             Supports Actions, Subjects, and Descriptors.
   * @see Action
   * @see Subject
   * @see Descriptor
   * @return The word passed in.
   *         For chaining and static assignment.
   */
  public static Word register(ResourceLocation key, Word word) {
    return register(key.toString(), word);
  }
}

File: common/src/main/kotlin/com/ssblur/scriptor/api/ScriptorRegistry.kt
package com.ssblur.scriptor.api

import com.ssblur.scriptor.api.word.Action
import com.ssblur.scriptor.api.word.Descriptor
import com.ssblur.scriptor.api.word.Subject
import com.ssblur.scriptor.api.word.Word
import com.ssblur.scriptor.registry.words.WordRegistry
import net.minecraft.resources.ResourceLocation

@Suppress("unused")
object ScriptorRegistry {
    /**
     * Registers a Word to Scriptor's WordRegistry, adding it to dictionaries.
     * @param key The key to register this word to.
     * Will overwrite any word of the same type already written to this key.
     * @param word The word to register.
     * Supports Actions, Subjects, and Descriptors.
     * @see Action
     * @see Subject
     * @see Descriptor
     * @return The word passed in.
     * For chaining and static assignment.
     */
    fun register(key: String, word: Word): Word {
        if (word is Action) return WordRegistry.register(key, word)!!
        if (word is Subject) return WordRegistry.register(key, word)!!
        if (word is Descriptor) return WordRegistry.register(key, word)!!
        throw RuntimeException("Unknown Word type! Expected one of Action, Subject, or Descriptor.")
    }

    /**
     * Registers a Word to Scriptor's WordRegistry, adding it to dictionaries.
     * @param key The key to register this word to.
     * Will overwrite any word of the same type already written to this key.
     * @param word The word to register.
     * Supports Actions, Subjects, and Descriptors.
     * @see Action
     * @see Subject
     * @see Descriptor
     * @return The word passed in.
     * For chaining and static assignment.
     */
    fun register(key: ResourceLocation, word: Word): Word {
        return register(key.toString(), word)
    }
}
